### ISSUE TITLE
hash pinning is network inefficient

### ISSUE THREAD

**Issue by skliarie** [ipfs/go-ipfs#7762]
Pinning of a CID with many nodes is not efficient.

For example, this CID has 3.3GB worth of data (1000 files, about 20000 nodes):
/ipfs/bafykbzacecwysablchj6hdqfa6u2iava3vexr57tgtpv4o23x34w2n4v2nuyy

When pinning it on otherwise idle ipfs node (about 20 KB/s idle traffic), it uses about 12GB of traffic (as seen by "ipfs stats bw") to transfer the the data completely.

IMHO there should be an optimization effort to group requests/replies for wanted blocks during transfer, with fallback to regular approach...

**Comment by skliarie** [comment]
Actually wait, let me confirm the issue on a different hash and make 100% there is no other ipfs ops in background, stay tuned.

**Comment by skliarie** [comment]
Did clean experiment. This hash (9.5GB) took 526m45.269s (8:46 hours) to download:
/ipfs/bafykbzacebsksnjlszf2534ujtvgzhmwymugnarhz5ebruik4dm6yprh23zou

No other pinning were in place at the time.
"ipfs stats bw" shows that during the time 57GB were downloaded.

Normally during "idle" 8 hours, only 2GB of traffic is downloaded. Yes, this much is because my node has public IP and participates in DHT.

**Comment by adlrocha** [comment]
IPFS uses Bitswap for pinning. The difference you see between the amount of "useful" data (9.5 GB in your last example), and the actual data exchanged through the wire (57GB) is due to the way Bitswap operates. You would see this same behaviour if instead of pinning you just `ipfs get` the CID.

Bitswap retrieves content "one block at a time" until it traverses the full DAG (it is not really one block at a time but "a bunch of blocks at a time"). This translates into a lot of roundtrip interactions with other peers in the network to first discover blocks for the content and then retrieving them. Furthermore, if you are exchanging a deep DAG, Bitswap needs to retrieve `level n` of the DAG in order to get the links of the next level before it can move on to `level n+1`, making parallelization difficult. All of this back and forth between your node and other seeders storing the content while traversing the DAG accounts for a lot of messages in a 9.8 GB, and is the reason for the overhead (if you test a 1:1 exchange between two IPFS nodes, one storing the content, and the other one fetching it you'll see a similar overhead).

In short, it doesn't seem to be that much a pinning inefficiency, but Bitswap's behaviour. We are already considering ways to improve this. If it is of your interest, I'd highly recommend checking the efforts and discussion flowing around this repo: https://github.com/protocol/beyond-bitswap

**Comment by aschmahmann** [comment]
closing to track in https://github.com/ipfs/go-ipfs/issues/8162
